Unstable unit test

```
[ERROR] Failures: 
[ERROR]   ExternalTypescriptSensorTest.should_log_error_if_no_typescript_found:268 
Expecting:
 <["No TypeScript compiler found in your project",
    "Global one referenced in 'NODE_PATH' will be used",
    "Analyzing 1 typescript file(s) with the following configuration file /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/tsconfig.json",
    "Starting external process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts`",
    "Error: Cannot find module 'typescript'",
    "External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts` returned an empty output. Run with -X for more information"]>
to contain:
 <["Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project"]>
but could not find:
 <["Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project"]>
```


```
[INFO] Running org.sonar.plugin.typescript.ExternalTypescriptSensorTest
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockEmptyResponse.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/foo/file.ts` returned an empty output. Run with -X for more information
ERROR Some error message
ERROR No tsconfig.json file found for /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/file.ts (looking up the directories tree). This file will not be analyzed.
ERROR Failed to find input file for path `some/path/file.ts`
ERROR Error: Cannot find module 'typescript'
ERROR Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts` returned an empty output. Run with -X for more information
ERROR Some error
ERROR 
ERROR /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classe...
ERROR throw errorMessage;
ERROR ^
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/rulesDeadlock.js` returned an empty output. Run with -X for more information
[ERROR] Tests run: 13, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.009 s <<< FAILURE! - in org.sonar.plugin.typescript.ExternalTypescriptSensorTest
[ERROR] should_log_when_empty_sonarts_out(org.sonar.plugin.typescript.ExternalTypescriptSensorTest)  Time elapsed: 0.077 s  <<< ERROR!
java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at org.sonar.plugin.typescript.ExternalTypescriptSensorTest.should_log_when_empty_sonarts_out(ExternalTypescriptSensorTest.java:187)
ERROR abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabc...
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ExternalTypescriptSensorTest.should_log_when_empty_sonarts_out:187 » IndexOutOfBounds

```

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java
@@ -257,7 +257,7 @@ public class ExternalTypescriptSensorTest {
     DefaultInputFile testInputFile = createTestInputFile(sensorContext, "dirWithoutTypeScript/file.ts");
 
     TestBundleFactory testBundle = new TestBundleFactory().command(node, resourceScript("/mockTypescriptNotFound.js"), testInputFile.absolutePath());
-    executeSensor(sensorContext, testBundle);
+    executeSensor(sensorContext, testBundle, new TestableErrorConsumer(500));
 
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("No TypeScript compiler found in your project");
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("Global one referenced in 'NODE_PATH' will be used");
@@ -293,10 +293,13 @@ public class ExternalTypescriptSensorTest {
   }
 
   private void executeSensor(SensorContextTester sensorContext, TestBundleFactory testBundle) {
-    SlowErrorConsumer slowErrorConsumer = new SlowErrorConsumer();
-    createSensor(testBundle, slowErrorConsumer).execute(sensorContext);
-    while (slowErrorConsumer.running != null && slowErrorConsumer.running) {
-      sleep();
+    executeSensor(sensorContext, testBundle, new TestableErrorConsumer(0));
+  }
+
+  private void executeSensor(SensorContextTester sensorContext, TestBundleFactory testBundle, TestableErrorConsumer errorConsumer) {
+    createSensor(testBundle, errorConsumer).execute(sensorContext);
+    while (errorConsumer.running != null && errorConsumer.running) {
+      sleep(errorConsumer.delay);
     }
   }
 
@@ -381,22 +384,27 @@ public class ExternalTypescriptSensorTest {
     }
   }
 
-  private static class SlowErrorConsumer extends ExternalProcessErrorConsumer {
+  private static class TestableErrorConsumer extends ExternalProcessErrorConsumer {
 
     private volatile Boolean running = null;
+    private int delay;
+
+    TestableErrorConsumer(int delay) {
+      this.delay = delay;
+    }
 
     @Override
     protected void readErrors(BufferedReader errorReader) {
       running = true;
-      sleep();
+      sleep(delay);
       super.readErrors(errorReader);
       running = false;
     }
   }
 
-  private static void sleep() {
+  private static void sleep(int millis) {
     try {
-      Thread.sleep(500);
+      Thread.sleep(millis);
     } catch (InterruptedException e) {
       Thread.currentThread().interrupt();
     }

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java
@@ -52,7 +52,6 @@ import org.sonar.api.utils.log.Loggers;
 import org.sonar.duplications.internal.pmd.TokensLine;
 import org.sonar.plugin.typescript.executable.ExecutableBundle;
 import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
-import org.sonar.plugin.typescript.executable.SonarTSCoreBundleFactory;
 import org.sonar.plugin.typescript.executable.SonarTSRunnerCommand;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -112,9 +111,8 @@ public class ExternalTypescriptSensorTest {
     SensorContextTester sensorContext = createSensorContext();
     DefaultInputFile testInputFile = createTestInputFile(sensorContext);
 
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), testInputFile.absolutePath()));
-
-    sensor.execute(sensorContext);
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), testInputFile.absolutePath());
+    executeSensor(sensorContext, bundleFactory);
 
     assertThat(sensorContext.allIssues()).hasSize(1);
 
@@ -152,9 +150,8 @@ public class ExternalTypescriptSensorTest {
     SensorContextTester sensorContext = createSensorContext();
     DefaultInputFile testInputFile = createTestInputFile(sensorContext);
 
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockSonarTSFileLevelIssue.js"), testInputFile.absolutePath()));
-
-    sensor.execute(sensorContext);
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockSonarTSFileLevelIssue.js"), testInputFile.absolutePath());
+    executeSensor(sensorContext, bundleFactory);
 
     assertThat(sensorContext.allIssues()).hasSize(1);
     Issue issue = sensorContext.allIssues().iterator().next();
@@ -166,8 +163,8 @@ public class ExternalTypescriptSensorTest {
     SensorContextTester sensorContext = createSensorContext();
     DefaultInputFile testInputFile = createTestInputFile(sensorContext);
 
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockSonarTSMissingData.js"), testInputFile.absolutePath()));
-    sensor.execute(sensorContext);
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockSonarTSMissingData.js"), testInputFile.absolutePath());
+    executeSensor(sensorContext, bundleFactory);
 
     assertThat(sensorContext.allIssues()).hasSize(0);
     assertThat(sensorContext.measure(testInputFile.key(), CoreMetrics.NCLOC).value()).isEqualTo(0);
@@ -179,15 +176,15 @@ public class ExternalTypescriptSensorTest {
     SensorContextTester sensorContext = createSensorContext();
     DefaultInputFile testInputFile = createTestInputFile(sensorContext);
 
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockEmptyResponse.js"), testInputFile.absolutePath()));
-
-    sensor.execute(sensorContext);
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockEmptyResponse.js"), testInputFile.absolutePath());
+    executeSensor(sensorContext, bundleFactory);
 
     assertThat(sensorContext.allIssues()).hasSize(0);
-    String errorMessage = logTester.logs(LoggerLevel.ERROR).get(1);
+    boolean containsMessage = logTester.logs(LoggerLevel.ERROR).stream().anyMatch(log -> log.contains("returned an empty output. Run with -X for more information"));
+    assertThat(containsMessage).isTrue();
 
-    assertThat(errorMessage).contains("returned an empty output. Run with -X for more information");
-    assertThat(logTester.logs().get(logTester.logs().size() - 1)).contains("foo/file.ts");
+    containsMessage = logTester.logs(LoggerLevel.ERROR).stream().anyMatch(log -> log.contains("foo/file.ts"));
+    assertThat(containsMessage).isTrue();
   }
 
   @Test
@@ -195,9 +192,8 @@ public class ExternalTypescriptSensorTest {
     SensorContextTester sensorContext = createSensorContext();
     DefaultInputFile testInputFile = createTestInputFile(sensorContext, "foo/bar/file.ts");
 
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), testInputFile.absolutePath()));
-
-    sensor.execute(sensorContext);
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), testInputFile.absolutePath());
+    executeSensor(sensorContext, bundleFactory);
 
     assertThat(sensorContext.allIssues()).hasSize(1);
   }
@@ -215,7 +211,7 @@ public class ExternalTypescriptSensorTest {
 
     SensorContextTester sensorContext = createSensorContext();
     createTestInputFile(sensorContext);
-    createSensor(testBundle).execute(sensorContext);
+    executeSensor(sensorContext, testBundle);
   }
 
   @Test
@@ -225,7 +221,7 @@ public class ExternalTypescriptSensorTest {
     DefaultInputFile inputFile = createTestInputFile(sensorContext, "file.ts");
     TestBundleFactory testBundle = new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), inputFile.absolutePath());
 
-    createSensor(testBundle).execute(sensorContext);
+    executeSensor(sensorContext, testBundle);
 
     assertThat(logTester.logs()).contains("No tsconfig.json file found for " + inputFile.absolutePath() + " (looking up the directories tree). This file will not be analyzed.");
   }
@@ -233,10 +229,10 @@ public class ExternalTypescriptSensorTest {
   @Test
   public void should_do_nothing_when_response_with_not_existing_file() throws Exception {
     new File(BASE_DIR, "not_exists.ts").getAbsolutePath();
-    ExternalTypescriptSensor sensor = createSensor(new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), "some/path/file.ts"));
+    TestBundleFactory bundleFactory = new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), "some/path/file.ts");
     SensorContextTester sensorContext = createSensorContext();
     createTestInputFile(sensorContext);
-    sensor.execute(sensorContext);
+    executeSensor(sensorContext, bundleFactory);
     assertThat(sensorContext.allIssues()).hasSize(0);
   }
 
@@ -247,7 +243,7 @@ public class ExternalTypescriptSensorTest {
     DefaultInputFile testInputFile = createTestInputFile(sensorContext, "dirWithoutTypeScript/file.ts");
 
     TestBundleFactory testBundle = new TestBundleFactory().command(node, resourceScript("/mockSonarTS.js"), testInputFile.absolutePath());
-    createSensor(testBundle).execute(sensorContext);
+    executeSensor(sensorContext, testBundle);
 
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("No TypeScript compiler found in your project");
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("Global one referenced in 'NODE_PATH' will be used");
@@ -261,7 +257,7 @@ public class ExternalTypescriptSensorTest {
     DefaultInputFile testInputFile = createTestInputFile(sensorContext, "dirWithoutTypeScript/file.ts");
 
     TestBundleFactory testBundle = new TestBundleFactory().command(node, resourceScript("/mockTypescriptNotFound.js"), testInputFile.absolutePath());
-    createSensor(testBundle, new ExternalProcessErrorConsumer()).execute(sensorContext);
+    executeSensor(sensorContext, testBundle);
 
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("No TypeScript compiler found in your project");
     assertThat(logTester.setLevel(LoggerLevel.DEBUG).logs()).contains("Global one referenced in 'NODE_PATH' will be used");
@@ -275,9 +271,7 @@ public class ExternalTypescriptSensorTest {
     DefaultInputFile testInputFile = createTestInputFile(sensorContext);
 
     TestBundleFactory testBundleFactory = new TestBundleFactory().command(node, resourceScript("/mockError.js"), testInputFile.absolutePath());
-    ExternalTypescriptSensor sensor = createSensor(testBundleFactory, new ExternalProcessErrorConsumer());
-
-    sensor.execute(sensorContext);
+    executeSensor(sensorContext, testBundleFactory);
     assertThat(logTester.setLevel(LoggerLevel.ERROR).logs()).contains("Some error");
     assertThat(sensorContext.allIssues()).hasSize(1);
   }
@@ -287,7 +281,7 @@ public class ExternalTypescriptSensorTest {
     TestBundleFactory testBundle = new TestBundleFactory().command(node, resourceScript("/rulesDeadlock.js"));
     SensorContextTester sensorContext = createSensorContext();
     createTestInputFile(sensorContext);
-    createSensor(testBundle).execute(sensorContext);
+    createSensor(testBundle, new LineTrimmingExternalProcessErrorConsumer()).execute(sensorContext);
 
     assertThat(sensorContext.allIssues()).hasSize(0);
   }
@@ -298,12 +292,16 @@ public class ExternalTypescriptSensorTest {
     return sensorContext;
   }
 
-  private ExternalTypescriptSensor createSensor() {
-    return createSensor(new TestBundleFactory());
+  private void executeSensor(SensorContextTester sensorContext, TestBundleFactory testBundle) {
+    SlowErrorConsumer slowErrorConsumer = new SlowErrorConsumer();
+    createSensor(testBundle, slowErrorConsumer).execute(sensorContext);
+    while (slowErrorConsumer.running != null && slowErrorConsumer.running) {
+      sleep();
+    }
   }
 
-  private ExternalTypescriptSensor createSensor(ExecutableBundleFactory executableBundleFactory) {
-    return createSensor(executableBundleFactory, new LineTrimmingExternalProcessErrorConsumer());
+  private ExternalTypescriptSensor createSensor() {
+    return createSensor(new TestBundleFactory(), new LineTrimmingExternalProcessErrorConsumer());
   }
 
   private ExternalTypescriptSensor createSensor(ExecutableBundleFactory executableBundleFactory, ExternalProcessErrorConsumer errorConsumer) {
@@ -383,4 +381,24 @@ public class ExternalTypescriptSensorTest {
     }
   }
 
+  private static class SlowErrorConsumer extends ExternalProcessErrorConsumer {
+
+    private volatile Boolean running = null;
+
+    @Override
+    protected void readErrors(BufferedReader errorReader) {
+      running = true;
+      sleep();
+      super.readErrors(errorReader);
+      running = false;
+    }
+  }
+
+  private static void sleep() {
+    try {
+      Thread.sleep(500);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
 }

